src/iso15118/message/common.cpp

```diff
@@ -95,22 +95,15 @@ float from_RationalNumber(const RationalNumber& in) {
 }
 
 RationalNumber from_float(float in) {
-
-    int8_t exponent = 0;
-
-    if (in - std::floor(in) != 0) {
-        exponent = 2;
-    }
-
-    for (;exponent > -4; exponent--) {
-        if ((in * pow(10, exponent)) < INT16_MAX) {
-            break;
-        }
-    }
-
     RationalNumber out;
-    out.exponent = (int8_t)(-exponent);
-    out.value = (int16_t)(in * pow(10, exponent));
+    if (in == 0.0) {
+        out.exponent = 0;
+        out.value = 0;
+        return out;
+    }
+    out.exponent = (int8_t)floor(log10(fabs(in)));
+    result.exponent -=3; // add 3 digits of precision
```

Comment:
`result` was not declared in this scope

Reply:
And this just artificially bumps everything to the "left", instead of looking for the ideal precision.

E.g. 211 maps to 2110 * 10^-1 instead of 211 * 10^0
0.211 maps to 2109 * 10^-4 instead of 211 * 10^-3

The loop did have some effect.

Reply:
Oops. I should have rebuilt it after that last edit.

Reply:
With as much as possible I figured the largest round number that comfortably fits in 16 bits, thus the 3 digits. 4 digits can overflow.